Add support of SMP. The third try

kolerov opened the pull request:
This set of commits introduces a support of SMP for ARCv2 and ARCv3 targets in QEMU. It's based on Cupertino's work. A partial support of ICI, IDU and GFRC subsystems is presented enough for running Linux.

Known issues:

* Some IDU commands like CMD_IDU_GEN_CIRQ, CMD_IDU_ACK_CIRQ
  and CMD_IDU_CHECK_FIRST are not supported yet. They may be
  implemented correctly only with a distinct interrupt controller for
  IDU. All devices are connected to the first core because of the same
  reason. It means that Linux cannot really control distributing of
  interrupts.
* Only reading operations for GFRC are supported.
* Only a small subset of ICI commands is supported.
* SMP for HS4x is slower than for HS5x and HS6x.

SMP mode may be turned on by -smp N option where N stands for a number of cores.

kolerov committed:
target/arc: v3: Use a distinct MMU state for each core

This patch moves data for MMU to core's state storage. It's
necessary for adding support of SMP since each core requires
it's own MMU state.

Signed-off-by: Yuriy Kolerov <ykolerov@synopsys.com>

kolerov committed:
target/arc: Add support of SMP

This commit introduces a support of SMP for ARCv2 and ARCv3 targets
in QEMU. A partial support of ICI, IDU and GFRC subsystems is presented
enough for running Linux.

Known issues:

* Some IDU commands like CMD_IDU_GEN_CIRQ, CMD_IDU_ACK_CIRQ
  and CMD_IDU_CHECK_FIRST are not supported yet. They may be
  implemented correctly only with a distinct interrupt controller for
  IDU. All devices are connected to the first core because of the same
  reason. It means that Linux cannot really control distributing of
  interrupts.
* Only reading operations for GFRC are supported.
* Only a small subset of ICI commands is supported.
* SMP for HS4x is slower than for HS5x and HS6x.

SMP mode may be turned on by -smp N option where N
stands for a number of cores.

Co-authored-by: Cupertino Miranda <cmiranda@synopsys.com>
Co-authored-by: Jose Abreu <jose.abreu@synopsys.com>
Signed-off-by: Yuriy Kolerov <ykolerov@synopsys.com>

claziss commented on target/arc/timer.c:
```diff
@@ -37,6 +37,12 @@
 #define CYCLES_TO_NS(VAL) (muldiv64(VAL, NANOSECONDS_PER_SECOND, FREQ_HZ))
 #define NS_TO_CYCLE(VAL)  (muldiv64(VAL, FREQ_HZ, NANOSECONDS_PER_SECOND))
 
+uint64_t get_global_cycles(void) {
```
Please add a short comment what this function does.

claziss commented on target/arc/arconnect.c:
```diff
@@ -17,115 +17,117 @@
 }
-static void arcon_status_set(CPUARCState *env, uint8_t status_core_id, uint8_t sender_core_id)
+
+static void arconnect_status_set(CPUARCState *env, uint8_t status_core_id,
```
Please add a comment about what this function  should do.

claziss commented on target/arc/arconnect.c:
```diff
@@ -134,7 +136,9 @@ static void arcon_status_set(CPUARCState *env, uint8_t status_core_id, uint8_t s
 }
-static void arcon_status_clr(CPUARCState *env, uint8_t status_core_id, uint8_t sender_core_id)
+
+static void arconnect_status_clr(CPUARCState *env, uint8_t status_core_id,
```
Please add a comment about what this function should do.

claziss commented on target/arc/arconnect.c:
```diff
@@ -143,7 +147,8 @@ static void arcon_status_clr(CPUARCState *env, uint8_t status_core_id, uint8_t s
 }
-static uint64_t arcon_status_read(CPUARCState *env, uint8_t core_id)
+
+static uint64_t arconnect_status_read(CPUARCState *env, uint8_t core_id)
```
Please add a comment about what this function should do.

claziss commented on target/arc/arconnect.c:
```diff
@@ -153,103 +158,178 @@ static uint64_t arcon_status_read(CPUARCState *env, uint8_t core_id)
-#define MCIP_IRQ 19
-
-static void arconnect_intercore_intr_unit_cmd(CPUARCState *env, enum arconnect_commands cmd, uint16_t param)
+static void arconnect_ici_cmd(CPUARCState *env, enum arconnect_commands cmd,
```
Please add a comment about what this function should do.

claziss commented on target/arc/arconnect.c:
```diff
@@ -153,103 +158,178 @@ static uint64_t arcon_status_read(CPUARCState *env, uint8_t core_id)
 
-#define ARCON_COMMAND(V) (V & 0xff)
-#define ARCON_PARAM(V) ((V >> 8) & 0xffff)
+static void arconnect_idu_cmd(CPUARCState *env, enum arconnect_commands cmd,
```
Please add a comment about what this function should do.

claziss commented on target/arc/arconnect.c:
```diff
@@ -153,103 +158,178 @@ static uint64_t arcon_status_read(CPUARCState *env, uint8_t core_id)
+    }
+}
+
+static void arconnect_gfrc_cmd(CPUARCState *env, enum arconnect_commands cmd,
```
Please add a comment about what this function should do.

claziss commented on target/arc/arconnect.c:
```diff
@@ -258,31 +338,31 @@ arconnect_regs_get(const struct arc_aux_reg_detail *aux_reg_detail, void *data)
-arconnect_regs_set(const struct arc_aux_reg_detail *aux_reg_detail,
-                   target_ulong val, void *data)
+
+void arconnect_regs_set(const struct arc_aux_reg_detail *aux_reg_detail,
```
Please add a comment about what this function should do.

claziss commented on target/arc/mmu-v6.c:
```diff
@@ -415,30 +347,76 @@ arc_mmuv6_aux_set(const struct arc_aux_reg_detail *aux_reg_detail,
 
 #define ALL1_64BIT (0xffffffffffffffff)
 
+/* TODO: This is for MMU48 only. */
```
Please be more verbose here.

kolerov replied:
Actually, I have a zero knowledge about MMU subsystem. I just rearranged some code blocks and substituted global MMU variables by local ones for each CPU. All comments are saved as is. MMU code must be revisited a bit later.

claziss commented on target/arc/mmu-v6.c:
```diff
@@ -415,30 +347,76 @@ arc_mmuv6_aux_set(const struct arc_aux_reg_detail *aux_reg_detail,
+
 static uint64_t
-root_ptr_for_vaddr(uint64_t vaddr, bool *valid)
+root_ptr_for_vaddr(CPUARCState *env, uint64_t vaddr, bool *valid)
```
Please add a comment about what this function should do.

kolerov replied:
The same as for the previous request.

claziss commented on target/arc/mmu-v6.c:
```diff
@@ -606,7 +584,7 @@ page_table_traverse(CPUARCState *env,
             }
 
-            mmu_fault_status = (l & 0x7);
+            env->mmu.v6.fault_status = (l & 0x7);
```
Can you use a different name here but not 'l' as it is difficult to see it/trace it.

abrodkin commented:
@kolerov could you please also add information on what is actually working with that implementation. I.e. if normally built images of Zephyr RTOS and the Linux kenrel for SMP HS3x/4x/5x/6x work as expected regardless of the mentioned known issues or there's a smaller subset of real software that works as of today.

kolerov commented:
> @kolerov could you please also add information on what is actually working with that implementation. I.e. if normally built images of Zephyr RTOS and the Linux kenrel for SMP HS3x/4x/5x/6x work as expected regardless of the mentioned known issues or there's a smaller subset of real software that works as of today.

I will explain it in details tomorrow. In short, Linux works well. I've not tested Zephyr yet.

kolerov commented:
I've update and simplified the implementation. Everything is documented now in details. Also, Inter-Core Interrupts unit if fixed - it was incorrect.

kolerov committed:
target/arc: Don't use a timer for RTC and enable it by default

Using QTimer for RTC implementation is redundent and meaningless
since it's just free-running clock.

Also, enable RTC by default. If it's not available then Linux
uses Timer 1 as fallback which is much less accurate.

Signed-off-by: Yuriy Kolerov <ykolerov@synopsys.com>

kolerov committed:
target/arc: Fix setting LF bit for spinlock

Signed-off-by: Yuriy Kolerov <ykolerov@synopsys.com>
